edgedb-client/src/builder.rs

```diff
@@ -322,6 +329,171 @@ impl fmt::Display for DisplayAddr<'_> {
     }
 }
 
+impl DSN {
+    fn new(dsn: &str) -> Result<Self, Error> {
+        let admin = dsn.starts_with("edgedbadmin://");
+        if !dsn.starts_with("edgedb://") && !admin {
+            return Err(ClientError::with_message(format!(
+                "String {:?} is not a valid DSN",
+                dsn,
+            )));
+        };
+        let url = url::Url::parse(dsn).map_err(|e| {
+            ClientError::with_source(e).context(format!("cannot parse DSN {:?}", dsn))
+        })?;
+        let query = url
+            .query_pairs()
+            .into_owned()
+            .into_iter()
+            .collect::<HashMap<_, _>>();
+        Ok(Self { url, admin, query })
+    }
+
+    async fn get_value<T>(
+        &mut self,
+        key: &'static str,
+        mut v: Option<T>,
+        conv: impl FnOnce(String) -> Result<T, Error>,
+        default: impl FnOnce() -> T,
+    ) -> Result<T, Error> {
+        let v_query = self.query.remove(key);
+        let k_env = format!("{key}_env");
+        let v_env = self.query.remove(k_env.as_str());
```

Comment:
```suggestion
        let v_env = self.query.remove(&k_env);
```
.. should be enough. `as_str` makes impression that you converting types here, which you actually don't do.